runtime/src/iree/builtins/ukernel/tools/e2e_matmul_benchmark.c

```diff
@@ -0,0 +1,366 @@
+  iree_uk_write_random_buffer(rowmajor_out_buffer, rowmajor_out_buffer_size,
+                              out_type, engine);
+  iree_uk_write_random_buffer(packed_lhs_buffer, packed_lhs_buffer_size,
+                              lhs_type, engine);
+  iree_uk_write_random_buffer(packed_rhs_buffer, packed_rhs_buffer_size,
+                              rhs_type, engine);
+  iree_uk_write_random_buffer(packed_out_buffer, packed_out_buffer_size,
+                              out_type, engine);
+  mmt4d_params.lhs_buffer = packed_lhs_buffer;
+  mmt4d_params.rhs_buffer = packed_rhs_buffer;
+  mmt4d_params.out_buffer = packed_out_buffer;
+  pack_lhs_params.in_buffer = rowmajor_lhs_buffer;
+  pack_lhs_params.out_buffer = packed_lhs_buffer;
+  pack_rhs_params.in_buffer = rowmajor_rhs_buffer;
+  pack_rhs_params.out_buffer = packed_rhs_buffer;
+  pack_out_params.in_buffer = rowmajor_init_out_buffer;
+  pack_out_params.out_buffer = packed_out_buffer;
+  unpack_out_params.in_buffer = packed_out_buffer;
+  unpack_out_params.out_buffer = rowmajor_out_buffer;
+
+  void* rowmajor_reference_out_buffer = malloc(rowmajor_out_buffer_size);
+  memcpy(rowmajor_reference_out_buffer, rowmajor_init_out_buffer,
+         rowmajor_out_buffer_size);
+  iree_uk_reference_rowmajor_matmul(
+      params->type, params->M, params->K, params->N, rowmajor_lhs_buffer,
+      rowmajor_rhs_buffer, rowmajor_reference_out_buffer, FLAG_accumulate);
+
+  int64_t batch_count = 1;
+  int64_t total_iterations = 0;
+  bool already_checked_results = false;
+  while (iree_benchmark_keep_running(benchmark_state, batch_count)) {
+    for (int i = 0; i < batch_count; ++i) {
+      iree_uk_pack(&pack_lhs_params);
+      iree_uk_pack(&pack_rhs_params);
+      if (FLAG_accumulate) iree_uk_pack(&pack_out_params);
+      iree_uk_mmt4d(&mmt4d_params);
+      iree_uk_unpack(&unpack_out_params);
+      if (!already_checked_results) {
+        already_checked_results = true;
+        if (memcmp(rowmajor_out_buffer, rowmajor_reference_out_buffer,
+                   rowmajor_out_buffer_size)) {
+          fprintf(stderr, "❌❌❌ Numerical error! ❌❌❌\n");
+          iree_abort();
+        }
+      }
+    }
+    total_iterations += batch_count;
+    batch_count *= 2;
+  }
+  iree_benchmark_set_items_processed(
+      benchmark_state,
+      total_iterations * 2 * params->M * params->N * params->K);
+  free(rowmajor_lhs_buffer);
+  free(rowmajor_rhs_buffer);
+  free(rowmajor_out_buffer);
+  free(rowmajor_init_out_buffer);
+  free(packed_lhs_buffer);
+  free(packed_rhs_buffer);
+  free(packed_out_buffer);
+  free(rowmajor_reference_out_buffer);
+  return iree_ok_status();
+}
+
+iree_uk_mmt4d_type_t iree_uk_mmt4d_type_parse(const char* type) {
+  if (!strcmp(type, "f32f32f32")) {
+    return iree_uk_mmt4d_type_f32f32f32;
+  }
+  if (!strcmp(type, "i8i8i32")) {
+    return iree_uk_mmt4d_type_i8i8i32;
+  }
+  fprintf(stderr, "Unhandled type: %s\n", type);
+  iree_abort();
+  return (iree_uk_mmt4d_type_t)0;
```

Comment:
Does the compiler complain about this ? (I'd have expected the abort would be treated as an unreachable here and so no return would be needed)

Reply:
The compiler seems happy enough about this, maybe as a product of our current set of warning flags, I don't know. I'm fine with omitting this too, but then the compiler would warn-error about this when `IREE_ATTRIBUTE_NORETURN` expands to empty -- I don't know if that includes any target that we care about.
https://sourcegraph.com/github.com/openxla/iree/-/blob/runtime/src/iree/base/attributes.h?L73-84